File: GeoBroker-Server/src/main/java/de/hasenburg/geobroker/server/distribution/BrokerArea.java
package de.hasenburg.geobroker.server.distribution;

import de.hasenburg.geobroker.commons.model.BrokerInfo;
import de.hasenburg.geobroker.commons.model.spatial.Geofence;
import de.hasenburg.geobroker.commons.model.spatial.Location;

import java.util.Objects;

// TODO migrate to commons and add serialization to kryo serializer
public class BrokerArea {

	private final BrokerInfo responsibleBroker;
	private final Geofence coveredArea;

	public BrokerArea(BrokerInfo responsibleBroker, Geofence coveredArea) {
		this.responsibleBroker = responsibleBroker;
		this.coveredArea = coveredArea;
	}

	BrokerInfo getResponsibleBroker() {
		return responsibleBroker;
	}

	Geofence getCoveredArea() {
		return coveredArea;
	}

	boolean CheckResponsibleBroker(String brokerId) {
		return responsibleBroker.getBrokerId().equals(brokerId);
	}

	boolean ContainsLocation(Location location) {
		return coveredArea.contains(location);
	}

	boolean intersects(Geofence messageGeofence) {
		return coveredArea.intersects(messageGeofence);
	}

	/*****************************************************************
	 * Generated methods
	 ****************************************************************/

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		BrokerArea that = (BrokerArea) o;
		return responsibleBroker.equals(that.responsibleBroker) && coveredArea.equals(that.coveredArea);
	}

	@Override
	public int hashCode() {
		return Objects.hash(responsibleBroker, coveredArea);
	}

}

File: GeoBroker-Commons/src/main/kotlin/de/hasenburg/geobroker/commons/model/disgb/BrokerArea.kt
package de.hasenburg.geobroker.commons.model.disgb

import de.hasenburg.geobroker.commons.model.spatial.Geofence
import de.hasenburg.geobroker.commons.model.spatial.Location
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonConfiguration

@Serializable
@SerialName("BrokerArea")
data class BrokerArea(val responsibleBroker: BrokerInfo, val coveredArea: Geofence) {

    fun hasResponsibleBroker(brokerId: String): Boolean {
        return responsibleBroker.brokerId == brokerId
    }

    fun containsLocation(location: Location?): Boolean {
        return coveredArea.contains(location)
    }

    fun intersects(messageGeofence: Geofence): Boolean {
        return coveredArea.intersects(messageGeofence)
    }

}

fun BrokerArea.toJson(): String {
    return Json.encodeToString(BrokerArea.serializer(), this)
}

/**
 * @throws [kotlinx.serialization.json.JsonDecodingException]
 */
fun String.toBrokerArea(): BrokerArea {
    return Json.decodeFromString(BrokerArea.serializer(), this)
}